plugins/example-output/dummy.c

```diff
@@ -0,0 +1,56 @@
+#include <stdio.h>
+#include "suricata-plugin.h"
+#include "dummy.h"
+
+#define OUTPUT_NAME "plugin:dummy"
+
+typedef struct Context_ {
+} Context;
+
+static int dummy_write(const char *buffer, int buffer_len, LogFileCtx *ctx) {
+    printf("dummy_write\n");
+    printf("dummy_write: rotate=%d\n", ctx->rotation_flag);
+    return 0;
+}
+
+static void dummy_close(LogFileCtx *ctx) {
+    printf("dummy_close\n");
+    SCFree(ctx->plugin);
+}
+
+static int dummy_open(LogFileCtx *ctx, ConfNode *conf) {
+    printf("conf: %p\n", conf);
+    Context *context = SCCalloc(1, sizeof(Context));
+    ctx->Write = dummy_write;
+    ctx->Close = dummy_close;
+    ctx->plugin = context;
+    return 0;
+}
+
+/**
+ * Called by Suricata to initialize the module. This module registers
+ * new file type to the JSON logger.
+ */
+void sc_plugin_init(void)
+{
+    PluginFileType *my_output = SCCalloc(1, sizeof(PluginFileType));
+    my_output->name = OUTPUT_NAME;
+    my_output->Open = dummy_open;
+    RegisterPluginFileType(my_output);
```

Comment:
Should this also register a free func?

Also, some plugins might have more complex shutdown requirements (e.g. properly close db session).  Should that be done in a 'Free', or should there be an additional 'Close'?

Reply:
I could see registering a destroy callback to associate the init function.  

Specifically to registering an output file type, it might make more sense to declare the Write/Open/Close callbacks during register, rather than in the open callback - which was just a bit of hack to limit modifications at this time. This will also help get rid of LogFileCtx in the open callback.

In this case Open would open the database connection, Close would close it.  And then a generic destroy would destroy any global context a plugin may have created.

Reply:
This is a good idea ... give the plugin a chance to finalize resources and any connections.

Reply:
I think in the end we'll have:
- Register
- Init
- ThreadInit
- Open
- Close
- ThreadDeInit
- Destroy

The thread init/deinit would be more clearly seen if something like the pcap logger were done as a plugin.Or multi-threaded JSON output.